docs/hugo/content/design/ADR-2023-02-Helm-Chart-Size-Limitations.md

```diff
@@ -0,0 +1,108 @@
+---
+title: '2023-01: Helm Chart Size Limitations and Workarounds'
+---
+
+## Context
+
+Helm v3 stores chart state in a Kubernetes secret by default. Kubernetes secrets are limited to a max size based on 
+etcd's configuration. The default etcd configuration limits secrets to a maximum size of 1MB.
+
+Our chart secret is fast approaching 1MB. In fact, 
+[outstanding PRs](https://github.com/Azure/azure-service-operator/pull/2727) already cause us to surpass the 1MB limit.
+
+Once this limit is reached, users will be unable to install our chart. When this happens, the error looks like:
+> Error: create: failed to create: Secret "sh.helm.release.v1.asov2.v1" is invalid: data: Too long: must have at most 1048576 bytes
+
+Digging in a bit, the Helm secret contains compressed data. This means that our chart itself can consume >1MB as long 
+as its compressed size is <1MB. Here's a breakdown of our chart size right now:
+
+| Scenario                         | Chart size | Helm secret size | Compression ratio |
+|----------------------------------|------------|------------------|-------------------|
+| v2.0.0-beta.4                    | 9917 KB    | 732 KB           | 0.07              |
+| v2.0.0-beta.4 (all docs removed) | 8600 KB    | 536 KB           | 0.06              |
+
+Unfortunately, Helm stores the chart _and all subcharts_ state in a single secret, which means that splitting the chart
+doesn't prevent us from bypassing this limit if we include the split-charts as subcharts.
+
+Helm supports alternative storage schemes such as a [SQL backend](https://helm.sh/docs/topics/advanced/#sql-storage-backend)
+that bypasses this limit, but we can't reasonably expect all users to use that.
+
+**We need to do something to overcome the 1MB max Helm chart limit**
+
+### Splitting the Helm Chart
+
+We could split the Helm chart into a number of separate charts, 1 for the operator like we have now, 
+and N charts containing just CRDs. The main question here is how to decide which CRDs go into which chart? There
+are a few options:
+1. Based on popularity/usage, with multiple groups together to limit the total number of charts
```

Comment:
Basing on popularity gets very political - some might perceive that resources not in the core chart were somehow less important, or not supported. Lots of room for misinterpretation.

Reply:
Agreed - I don't love it either.